### PR TITLE
ready for review

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2019-10-08-tf2-whatchanges/tf2.Rmd [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+    img = tf$image$convert_image_dtype(.x$img, dtype = tf$float32),
+    mask = tf$image$convert_image_dtype(.x$mask, dtype = tf$float32)
+  )) %>% 
+  dataset_map(~.x %>% list_modify(
+    img = tf$image$resize(.x$img, size = shape(128, 128)),
+    mask = tf$image$resize(.x$mask, size = shape(128, 128))
+  ))
+```
+
+Note how once you know what these functions do, they free you of a lot of thinking (remember how in the "old" Keras approach to image preprocessing, you were doing things like dividing pixel values by 255 "by hand"?)
```
Comment (batpigandme) [inline _posts/2019-10-08-tf2-whatchanges/tf2.Rmd]:
```suggestion
Note how once you know what these functions do, they free you from a lot of thinking (remember how in the "old" Keras approach to image preprocessing, you were doing things like dividing pixel values by 255 "by hand"?)
```

File: _posts/2019-10-08-tf2-whatchanges/tf2.Rmd [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+  step_bucketized_column(age, boundaries = c(18, 25, 30, 35, 40, 45, 50, 55, 60, 65)) %>% 
+  step_indicator_column(thal) %>% 
+  step_embedding_column(thal, dimension = 2) %>% 
+  step_crossed_column(c(thal, bucketized_age), hash_bucket_size = 10) %>%
+  step_indicator_column(crossed_thal_bucketized_age)
+
+spec %>% fit()
+```
+
+What happened here is that we told TensorFlow, please take all numeric columns (besides a few ones listed exprès) and scale them; take column `thal`, treat it as categorical and create an embedding for it; discretize `age` according to the given ranges; and finally, create a _crossed column_ to capture interaction between `thal` and that discretized age-range column. ^[`step_indicator_column` is there (twice) for technical reasons. Our [post on feature columns](https://blogs.rstudio.com/tensorflow/posts/2019-07-09-feature-columns/) explains.]
```
Comment (batpigandme) [inline _posts/2019-10-08-tf2-whatchanges/tf2.Rmd]: This is a really nice translation of code into words 😀

File: _posts/2019-10-08-tf2-whatchanges/tf2.Rmd [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+
+This is nice, but when creating the model, we'll still have to define all those layers, right? (Which should be pretty cumbersome, having to figure out all the right dimensions...)
```
Comment (batpigandme) [inline _posts/2019-10-08-tf2-whatchanges/tf2.Rmd]:
```suggestion
This is nice, but when creating the model, we'll still have to define all those layers, right? (Which would be pretty cumbersome, having to figure out all the right dimensions...)
```

File: _posts/2019-10-08-tf2-whatchanges/tf2.Rmd [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+input <- layer_input_from_dataset(hearts %>% select(-target))
+
+output <- input %>% 
+  layer_dense_features(feature_columns = dense_features(spec)) %>% 
+  layer_dense(units = 1, activation = "sigmoid")
+```
+
+From then on, it's just normal `keras` `compile` and `fit`. See the [vignette](https://tensorflow.rstudio.com/tools/tfdatasets/articles/feature_columns.html) for the complete example. There also is a [post on feature columns](https://blogs.rstudio.com/tensorflow/posts/2019-07-09-feature-columns/) explaining more of how this works, and illustrating the time-and-nerve-saving effect by comparing with the pre-feature-spec way of working with heterogeneous datasets.
+
+As a last item on the topics of preprocessing and feature engineering, let's look at a much-promising thing to come in what we hope is the near future.
```
Comment (batpigandme) [inline _posts/2019-10-08-tf2-whatchanges/tf2.Rmd]:
```suggestion
As a last item on the topics of preprocessing and feature engineering, let's look at a promising thing to come in what we hope is the near future.
```

File: _posts/2019-10-08-tf2-whatchanges/tf2.Rmd [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+
+### Keras preprocessing layers
+
+Reading what above, we wrote about using `tfdatasets` for building a input pipeline, and seeing how we gave an image loading example, you may have been wondering: What about data augmentation functionality available, historically, through `keras`? Like `image_data_generator`?
```
Comment (batpigandme) [inline _posts/2019-10-08-tf2-whatchanges/tf2.Rmd]:
```suggestion
Reading what we wrote above about using `tfdatasets` for building a input pipeline, and seeing how we gave an image loading example, you may have been wondering: What about data augmentation functionality available, historically, through `keras`? Like `image_data_generator`?
```

File: _posts/2019-10-08-tf2-whatchanges/tf2.Rmd [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+  ) %>%
+  step_numeric_column(Age, Fee, Quantity, normalizer_fn = scaler_standard()) %>%
+  step_categorical_column_with_vocabulary_list(
+    has_type("string"), -Description, -RescuerID, -img_path, -PetID, -Name
+  ) %>%
+  step_embedding_column(Breed1:Health, State)
+
+```
+
+Both usage modes illustrate the high potential of working with Hub modules. Just be cautioned that as of today, not every model published will work with TF 2.
```
Comment (batpigandme) [inline _posts/2019-10-08-tf2-whatchanges/tf2.Rmd]:
```suggestion
Both usage modes illustrate the high potential of working with Hub modules. Just be cautioned that, as of today, not every model published will work with TF 2.
```

File: _posts/2019-10-08-tf2-whatchanges/tf2.Rmd [hunk 7]
```diff
@@ -0,0 +1,414 @@
+---
+title: "TensorFlow 2.0 is there - what changes for R users?"
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2019-10-08-tf2-whatchanges/tf2.Rmd]: might be a "false friend" (from German :-)) thanks!

File: _posts/2019-10-08-tf2-whatchanges/tf2.Rmd [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+description: > 
+  TensorFlow 2.0 finally having been released last week, as R users we have two kinds of questions. First, will my keras code still run? And second, what is it that changes? In this post, we answer both and then, give a tour of exciting new developments in the r-tensorflow ecosystem.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2019-10-08-tf2-whatchanges/tf2.Rmd]: not at all, I'll totally do like you say :-)
(plus, I know I tend to overdo things in that respect)
(((I think it was called "gerund", but ... too many lives ago ;-)))

File: _posts/2019-10-08-tf2-whatchanges/tf2.Rmd [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+## Don't panic
+
+- If you are using `keras` in standard ways, such as those depicted in most code examples and tutorials seen on the web, and things have been working fine for you in recent `keras` releases (>= 2.2.4.1), don't worry. Most everything should work without major changes.
+- If you are using an older release of `keras` (< 2.2.4.1), syntactically things should work fine as well, but you will want to check for changes in behavior/performance.
+
+And now for some news and background. This post aims to do three things:
+
+- Explain the above _all-clear_ statement. Is it really that simple -- what exactly is going on?
+- Characterize the changes brought about by TF 2, from the _point of view of the R user_.
+- And perhaps most interestingly: Take a look at what is going on, in the `r-tensorflow` ecosystem, around new functionality brought about, or related to somehow, the advent of TF 2.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2019-10-08-tf2-whatchanges/tf2.Rmd]: makes sense :-)

**Comment by dfalbel** [comment]
LGTM! :)

**Comment by skeydan** [comment]
Thanks a lot @batpigandme & @dfalbel !!
